### PR TITLE
runtime: Best-effort one T::Types::Simple instance per module & one T.nilable(<mod>) per module

### PR DESCRIPTION
**Pull request by djudd-stripe**
The basic idea here is:
1. For `T::Types::Simple`, we cache an instance on the module that we wrap, as an instance variable. We prefer to use this instance rather than construct a new one when using standard helpers.
2. For `T::Types::Union` uses that are a simple union of NilClass and a T::Types::Simple (ie, a simple T.nilable use), we cache an instance on the corresponding (cached) T::Types::Simple instance. We prefer to use this instance rather than construct a new one when using standard helpers.

I also added simple microbenchmarks for runtime typechecking.

Notably this does not _guarantee_ that only one equivalent instance of a T::Types::Simple or simple-nilable T::Union can exist. There are a couple of reasons that's difficult:
1. For T::Types::Simple, modules can be frozen. So we'd have to do the caching in something like a global Hash keyed by module instead. That'd be a little more complex and also I worry that would perform worse due to worse memory locality, although I admit I haven't tested.
2. For T::Types::Union, catching all possible ways that a simple-nilable union could define would require running after the `flat_map` and `uniq` logic currently in `initialize`, and I couldn't find a way of doing that which didn't reverse any gains in the benchmarks.

Since `new` on `T::Types::Simple` and `T::Types::Union` is public, for backwards compatibility we'd also have to override `new` instead of just adding a wrapper. That's doable but seems smellier.

I could be persuaded to reconsider the above reasoning and try to actually offer a uniqueness guarantee. It'd have the advantage of giving us cheaper equality and hash implementations. But those aren't hot paths, are they?

Another alternative would be to just hardcode constants for a handful of the most commonly-wrapped classes, like Integer, String, Float, Symbol etc. That seems like it would end up with less benefit for just as much complexity, though.

### Motivation
This is aiming to help with two things:
1. Total memory usage from `sig` definitions in a large codebase;
2. Runtime allocations from `T.let` uses (where type definitions aren't necessarily restricted to load-time like they are for `sig`).

I don't have memory usage data currently, but benchmarks before:
```
T.let(..., Integer) twice, μs/iter
  0.930970   0.000963   0.931933 (  0.932905)
sig {params(x: Integer).void} twice, μs/iter
  1.084934   0.001938   1.086872 (  1.087867)
T.let(..., T.nilable(Integer)) twice, μs/iter
  5.485798   0.003305   5.489103 (  5.491234)
sig {params(x: T.nilable(Integer)).void} twice, μs/iter
  1.981453   0.000917   1.982370 (  1.983042)
T.let(..., Example) twice, μs/iter
  0.814551   0.000244   0.814795 (  0.814936)
sig {params(x: Example).void} twice, μs/iter
  0.948986   0.000253   0.949239 (  0.949511)
T.let(..., T.nilable(Example)) twice, μs/iter
  5.453973   0.003054   5.457027 (  5.458655)
sig {params(x: T.nilable(Example)).void} twice, μs/iter
  2.053776   0.001400   2.055176 (  2.055702)
```

... vs after:
```
T.let(..., Integer) twice, μs/iter
  0.723203   0.000327   0.723530 (  0.723655)
sig {params(x: Integer).void} twice, μs/iter
  0.933413   0.001082   0.934495 (  0.934596)
T.let(..., T.nilable(Integer)) twice, μs/iter
  2.315996   0.000328   2.316324 (  2.316441)
sig {params(x: T.nilable(Integer)).void} twice, μs/iter
  1.862074   0.000523   1.862597 (  1.862729)
T.let(..., Example) twice, μs/iter
  0.704537   0.000083   0.704620 (  0.704666)
sig {params(x: Example).void} twice, μs/iter
  0.939468   0.000371   0.939839 (  0.940009)
T.let(..., T.nilable(Example)) twice, μs/iter
  2.398132   0.000883   2.399015 (  2.399266)
sig {params(x: T.nilable(Example)).void} twice, μs/iter
  1.943312   0.001264   1.944576 (  1.945432)
```

### Test plan
Existing tests plus benchmarks. Will also test in pay-server before merging.

### CODE REVIEW

File: gems/sorbet-runtime/lib/types/types/simple.rb [hunk 1]
```diff
@@ -29,5 +29,27 @@ def valid?(obj)
         false
       end
     end
+
+    def to_nilable
+      @nilable ||= T::Types::Union.new([self, T::Utils::Nilable::NIL_TYPE])
```
Comment (jez) [inline gems/sorbet-runtime/lib/types/types/simple.rb:34]: Wow I did not know we already had `NIL_TYPE` stored in a constant

**Comment by jez** [comment]
> It'd have the advantage of giving us cheaper equality and hash implementations. But those aren't hot paths, are they?

No, I don't think we ever check structural equality of types.

File: gems/sorbet-runtime/lib/types/types/simple.rb [hunk 2]
```diff
@@ -29,5 +29,27 @@ def valid?(obj)
...
+    end
+
+    module Private
+      module Pool
+        def self.type_for_module(mod)
+          cached = mod.instance_variable_get(:@__as_sorbet_simple_type)
+          return cached if cached
+
+          type = Simple.unpooled(mod)
+          mod.instance_variable_set(:@__as_sorbet_simple_type, type) unless mod.frozen?
```
Comment (jez) [inline gems/sorbet-runtime/lib/types/types/simple.rb:49]: Can you add a test using frozen modules?

Specifically:

- That it doesn't raise when calling this method
- That the resulting types are not reference equal

**Comment by djudd-stripe** [comment]
I think I can add a test for `reference equality as a replacement for structural equality`. Agreed re the mutability question but can't think of a simple mitigation. Fortunately the temptation to mutate the type is probably small?

**Comment by djudd-stripe** [comment]
Added tests as requested

**Comment by djudd-stripe** [comment]
pay-server build looks good (a few failing tests but they look related to using a non-master version of sorbet-runtime at all) and this build is passing now, so ready for re-review whenever @jez

**Comment by bbuchalter** [comment]
This change seems to be incompatible with assumptions made by ActiveSupport. 

When calling [`as_json`](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/object/json.rb#L57) on an object, [#instance_values](https://apidock.com/rails/v6.0.0/Object/instance_values) is invoked, which transforms instance variables on that object into a hash. 

With the introduction of [`mod.instance_variable_set(:@__as_sorbet_simple_type, type) unless mod.frozen?`](https://github.com/sorbet/sorbet/pull/3309/files#diff-65c74fbebfb10aa05f9592ed685c8561R44), when calling `as_json` on a class, we trigger infinite recursion and a `SystemStackError: stack level too deep` exception.

Example:
```
[57] pry(#<Hash>)> Data.as_json
SystemStackError: stack level too deep
[58] pry(#<Hash>)> Data.instance_values.fetch('__as_sorbet_simple_type')
=> #<T::Types::Simple:0x00007fcee2d17e38 @raw_type=Data>
```

**Comment by djudd-stripe** [comment]
What's the context in which you want to call `as_json` on a class?

It seems like the fix here is probably patching `instance_values` to ignore this variable; does that make sense? Maybe in sorbet-rails or similar?

**Comment by bbuchalter** [comment]
> What's the context in which you want to call as_json on a class?

I'm not saying this is a normal pattern, or even a good idea, but programmers gonna program...

> It seems like the fix here is probably patching instance_values to ignore this variable; does that make sense? Maybe in sorbet-rails or similar?

Infinity agree. I can open a PR if that would be helpful.

**Comment by djudd-stripe** [comment]
Well, if sorbet-rails is the right place, we don't maintain that, but a PR seems like a reasonable next step to me!

**Pull request merged**